### PR TITLE
Allow leading CR/LF in HTTP request lines in Kestrel

### LINKED ISSUE #40747: HttpRequest with leading \r before request-line rejected with BadRequestException
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

HttpRequest with leading \r before request-line rejected with BadRequestException

### Expected Behavior

Should support (simply skip) similar to IIS or other server implementation.

https://stackoverflow.com/questions/46924558/http-leading-blank-line-in-header-a-valid-http-request-header

### Steps To Reproduce

```
00000000  0d 50 4f 53 54 20 2f 50  6f 73 74 2f  .POST /Post/
```

By using socket to simulate the leading \r before request-line.

```c#
var url = new Uri(args[0]);
int hostPort = 80;
var entry = await Dns.GetHostEntryAsync(url.Host);
var hostIp = entry.AddressList.FirstOrDefault();
var hostep = new IPEndPoint(hostIp, hostPort);
using (var sock = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
{
    await sock.ConnectAsync(hostep);
    var strb = new StringBuilder();
    // pass leading \r
    strb.Append($"\rGET / HTTP/1.1\r\n");
    strb.Append($"HOST: {url.Host}\r\n");
    strb.Append($"Connection: close\r\n");
    strb.Append($"User-Agent: testclient/1.1\r\n");
    strb.Append("\r\n");

    Console.WriteLine(strb);
    sock.Send(Encoding.UTF8.GetBytes(strb.ToString()));

    var buf = new byte[4096];
    var bytes = sock.Receive(buf, buf.Length, 0);
    Console.WriteLine(Encoding.ASCII.GetString(buf, 0, bytes));
    sock.Close();
}
```

### Exceptions (if any)

```c#
krpbfe: Received a bad request from client IP ::ffff:205.154.150.223 port 9073 for sni , error: Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Invalid request line: ''
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.GetUnknownMethodLength(ReadOnlySpan`1 span)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, ReadOnlySpan`1 requestLine)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, SequenceReader`1& reader)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TakeStartLine(SequenceReader`1& reader)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ParseRequest(SequenceReader`1& reader)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
```

### .NET Version

6.0.102

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by adityamandaleeka**
Fix https://github.com/dotnet/aspnetcore/issues/40747

Changing this in main first, will backport to 6 after review/merge.

### CODE REVIEW

File: src/Servers/Kestrel/Core/test/StartLineTests.cs [hunk 1]
```diff
@@ -516,6 +516,27 @@ public void AuthorityForms(string rawTarget, string path, string query)
...
+    public void LeadingCrLfAreAllowed(string startOfRequestLine)
+    {
+        var rawTarget = "http://localhost/path1?q=123&w=xyzw";
+        Http1Connection.Reset();
+        // RawTarget, Path, QueryString are null after reset
+        Assert.Null(Http1Connection.RawTarget);
+        Assert.Null(Http1Connection.Path);
+        Assert.Null(Http1Connection.QueryString);
+
+        var ros = new ReadOnlySequence<byte>(Encoding.ASCII.GetBytes($"{startOfRequestLine}CONNECT {rawTarget} HTTP/1.1\r\n"));
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/test/StartLineTests.cs:535]: Only connect?

Comment (adityamandaleeka) [reply src/Servers/Kestrel/Core/test/StartLineTests.cs:535]: The method doesn't matter for this today but I suppose we could somehow change that in the future :smile:. I'll vary the method.

Comment (davidfowl) [reply src/Servers/Kestrel/Core/test/StartLineTests.cs:535]: I know I guess I'd like to see it added for all the verbs we hardcode and then one unknown

**Review by Tratcher — APPROVED** [review]
Wow, that was easier than I expected.

**Comment by davidfowl** [comment]
Run the benchmarks before merging so we understand the impact.

cc @sebastienros PR benchmarks?

**Review by halter73 — APPROVED** [review]

**Comment by davidfowl** [comment]
I think we should do a quick branch to see if the line starts with \r or n before calling `AdvancePastAny` and doing another performance run,

**Comment by davidfowl** [comment]
@sebastienros When we do comparison runs can we choose a fixed number of requests?

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 2]
```diff
@@ -55,6 +55,9 @@ public HttpParser(bool showErrorDetails)
     /// </summary>
     public bool ParseRequestLine(TRequestHandler handler, ref SequenceReader<byte> reader)
     {
+        // Skip any leading \r or \n on the request line. This is not technically allowed, but
+        // but apparently there are enough clients relying on this that it's worth allowing.
+        reader.AdvancePastAny((byte)'\r', (byte)'\n');
```
[the remark this answers is not in the document]
Comment (adityamandaleeka) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs]: Thanks, fixed in latest commit.

**Comment by adityamandaleeka** [comment]
I'm not convinced we'll see anything in the benchmark results, but given that the TryPeek is super quick and aggressively inlined, and it should be a tiny bit quicker for the overwhelmingly common case where these characters are not present, I'm happy with this approach.

**Review by JamesNK — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 3]
```diff
@@ -55,6 +55,14 @@ public HttpParser(bool showErrorDetails)
     /// </summary>
     public bool ParseRequestLine(TRequestHandler handler, ref SequenceReader<byte> reader)
     {
+        // Skip any leading \r or \n on the request line. This is not technically allowed, but
```
Comment (TheBlueSky) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:58]:
```suggestion
        // Skip any leading \r or \n on the request line. This is not technically allowed,
```

**Pull request merged**